aip/0194.md

```diff
@@ -0,0 +1,237 @@
+    // It already exists.
+    resp = client.Get("Foo")
+  } else {
+    // An actual error.
+    panic(err)
+  }
+}
+_ = resp // TODO: Use resp.
+```
+
+The above code is quite natural and exists in many codebases already: it is the
+easy answer to asynchrony within a system, and a cheaper/more performant way
+of writing code that doesn't know about state (the alternative is to always
+perform a `READ` before a `CREATE`/`DELETE`, which is superfluous).
+
+In comparison to the relatively small downside, there is a large advantage.
+Users will _almost always_ retry `INTERNAL` and `UNAVAILABLE`. By building it
+into a client library instead, users are saved from implementing retry loops,
+deadlines, exponential backoff, and jitter.
+
+Finally, consider that most users _will have to write the if statement either
+way_. After all, a user who implements retry logic themselves will _still_ run
+into the same problem (and solution) described above.
+
+#### Absolute correctness
+
+There is an argument that some user may need to be absolutely sure of the
+state of a system before retrying a non-`GET` request. Such a user is an
+argument in favour of only retrying `GET` requests; never retrying `UPDATE`,
+`CREATE`, or `DELETE`. However, this set of users likely has very niche
+requirements, and is dwarfed by the set of users for whom retrying is
+advantageous.
+
+A simple answer exists to support this set of users: provide an opt-out for
+retries. This approach fits the generally-accepted goal of client libraries
+striving for setting good defaults with opt-outs for the minority of users.
+
+### The concurrency fallacy
+
+Another argument against retrying `CREATE` and `DELETE` is as such:
+
+```
+Thread1: Get("Foo")
+Thread1: > NOT_FOUND
+Thread2: Get("Foo")
+Thread2: > NOT_FOUND
+Thread1: Create("Foo") // Let's work under the assumption that this should be retried.
+Thread1: // ----> UNAVAILABLE
+Thread2: Create("Foo")
+Thread2: // ----> OK
+Thread1: // ----> Create("Foo") # retry #1
+Thread1: // ----> ALREADY_EXISTS
+Thread1: > ALREADY_EXISTS
+```
+
+Here, a system has two threads that convince themselves that "Foo" does not
+exist, and then try to create it. Because Thread1 retried the first failed
+Create request, it unexpectedly returned `ALREADY_EXISTS`.
+
+However, it is easy to see that _one of_ these requests would have to receive
+`ALREADY_EXISTS`, even without retries. So, the problem here is not retrying:
+the problem is that there is no synchronization between two components of the
+system that are both concerned with the same resource. Removing the retry logic
+would not solve the problem.
+
+There are several solutions to this problem that do not involve retrying:
+
+- Run these requests in a read-write transaction, ensuring atomicity.
+- Make the two components in the system synchronous.
+- Add a piece of code that considers `ALREADY_EXISTS` a success case (as above
+  described in "Idempotency and pragmatism").
+
```

Comment:
Add a `## Further reading` section that points to AIP-154, and add a further reading section to that AIP that points here.

Reply:
Done